### PR TITLE
[fix](memory) Fix metadata memory tracking and profile

### PR DESCRIPTION
**Pull request by xinyiZzz**
### What problem does this PR solve?

Problem Summary:

1. `MemTrackerLimiter` and `MemoryProfile` add metadata and cache type.
2. metadata memory consists of rowsets, segments, and tablets.
3. `all_segments_mem_tracker` excludes the segments in SegmentCache, and all segments use the same memory calculation method.
4. segments memory is estimated, which is not accurate and can only be used as a reference.
5. rowsets and tablets only count part of the memory, mainly PB.

TODO, Metadata also has some memory that can be allocated using Doris Allocator.

```
MemoryProfile:
  MemoryOverviewSnapshot:
     - PhysicalMemory(VmRSS) Current: 5.85 GB (Peak: 7.13 GB)
     - VirtualMemory(VmSize) Current: 61.35 GB (Peak: 61.67 GB)
    UntrackedMemory:
       - Memory Current: 3.61 GB (Peak: 3.77 GB)
    TrackedMemory:
       - Memory Current: 2.24 GB (Peak: 3.48 GB)
      TasksMemory:
         - Memory Current: 0 (Peak: 749.17 MB)
           - ReservedMemory Current: 0 (Peak: 0)
        Details:
           - Compaction Current: 0 (Peak: 13.60 MB)
           - Load Current: 0 (Peak: 749.17 MB)
             - AllMemTablesMemory Current: 0 (Peak: 853.20 MB)
           - Other Current: 0 (Peak: 0)
           - Query Current: 0 (Peak: 65.20 MB)
           - SchemaChange Current: 0 (Peak: 0)
      GlobalMemory:
         - Memory Current: 55.03 MB (Peak: 70.39 MB)
      MetadataMemory:
         - Memory Current: 972.13 MB (Peak: 972.13 MB)
      CacheMemory:
         - Memory Current: 360.21 MB (Peak: 675.32 MB)
      JemallocMemory:
         - Memory Current: 905.55 MB (Peak: 1.80 GB)
        Details:
           - Cache Current: 646.47 MB (Peak: 1.59 GB)
           - Metadata Current: 259.08 MB (Peak: 259.72 MB)
  GlobalMemorySnapshot:
    Orphan@global@id=3144f940fa84f930-d82d459289789494:
       - Memory Current: 1.73 MB (Peak: 1.73 MB)
    IOBufBlockMemory@global@id=1f4316bc5fb4c3d4-73f14d3652ed269b:
       - Memory Current: 2.65 GB (Peak: 2.65 GB)
    PointQueryExecutor@global@id=794878deaa4d8f56-769fbc87ccb7ec8d:
       - Memory Current: 36.90 MB (Peak: 36.90 MB)
    BlockCompression@global@id=0a4d429b1915dd8b-c2e429b1186d69a8:
       - Memory Current: 563.01 MB (Peak: 563.01 MB)
    RowIdStorageReader@global@id=a546ca5878c9204e-c4c72ea9088ca9a0:
       - Memory Current: 101.25 KB (Peak: 101.25 KB)
    SubcolumnsTree@global@id=294bde93c01d4ba6-4ce7b3290489a1b8:
       - Memory Current: 95.61 MB (Peak: 95.61 MB)
    S3FileBuffer@global@id=bc437d624d9a234d-03eb4bbc3d030a99:
       - Memory Current: 0 (Peak: 0)
  MetadataMemorySnapshot:
    Tablets(not in SchemaCache, TabletSchemaCache)@metadata@id=7646d676ad9f8de9-98beb5f1a8aa739b:
       - Memory Current: 26.84 GB (Peak: 26.84 GB)
    Segments(not in SegmentCache)@metadata@id=864b4887722b327a-516c912d4602b59b:
       - Memory Current: 4.98 MB (Peak: 4.98 MB)
    Rowsets@metadata@id=f0440aa247730497-752a527c71e3c393:
       - Memory Current: 5.57 GB (Peak: 5.57 GB)
    SegmentCache[size]@metadata@id=b34e798db2c25109-04c7ff3c722e4cbb:
       - Memory Current: 4.86 GB (Peak: 4.86 GB)
    SchemaCache[number]@metadata@id=9f41f61531da3fd9-163834a6a6f83eba:
       - Memory Current: 103.38 MB (Peak: 103.38 MB)
    TabletSchemaCache[number]@metadata@id=f9475c16ba7ea1ab-1d28b38568ef0b8d:
       - Memory Current: 170.65 MB (Peak: 170.65 MB)
  CacheMemorySnapshot:
    QueryCache@cache@id=504af426a68103cf-a60f36a48bc1998a:
       - Memory Current: 0 (Peak: 0)
    DataPageCache[size]@cache@id=f744bb4ef53c307b-4c021a8f9b8a77a5:
       - Memory Current: 14.33 GB (Peak: 14.33 GB)
    IndexPageCache[size]@cache@id=cb4ed0e456f081d4-d635edda9aa2778b:
       - Memory Current: 5.25 GB (Peak: 5.25 GB)
    PKIndexPageCache[size]@cache@id=8e4bf12b4d39133a-e0d878a18d73619b:
       - Memory Current: 2.72 GB (Peak: 2.72 GB)
    PointQueryRowCache[size]@cache@id=024dc624c80205f1-ca3221a154e4b68a:
       - Memory Current: 94.84 KB (Peak: 94.84 KB)
    CommonObjLRUCache[number]@cache@id=af40d4a03b391540-43338381c8e826a4:
       - Memory Current: 0 (Peak: 0)
    PointQueryLookupConnectionCache[number]@cache@id=284fa848dd97ccc8-56c61386a4950784:
       - Memory Current: 337.97 KB (Peak: 337.97 KB)
    InvertedIndexSearcherCache[size]@cache@id=f5404416ad0a9465-3cf12f5224276d8a:
       - Memory Current: 111.52 MB (Peak: 111.52 MB)
    InvertedIndexQueryCache[size]@cache@id=6f4d3c55508e3531-3fafe328c6d3e1ab:
       - Memory Current: 51.97 MB (Peak: 51.97 MB)
    QueryCache[size]@cache@id=274d47670e0796ad-542ad958334ff988:
       - Memory Current: 0 (Peak: 0)
    LastSuccessChannelCache[size]@cache@id=08449badfd6abb81-5b30599bfc60f79f:
       - Memory Current: 0 (Peak: 0)
    TabletColumnObjectPool[number]@cache@id=7b42da8391945265-06697f83a6bf01a7:
       - Memory Current: 151.67 MB (Peak: 151.67 MB)
    MowTabletVersionCache[number]@cache@id=f345a1adbd2a87dd-74b9caa3b30cf18b:
       - Memory Current: 150.94 MB (Peak: 150.94 MB)
    CreateTabletRRIdxCache[number]@cache@id=f043eee91e9aef2d-ef3a2245d50545ae:
       - Memory Current: 15.20 MB (Peak: 15.20 MB)
    MowDeleteBitmapAggCache[size]@cache@id=df47a8ed5bf3f11f-ac944d7d7e79d4b3:
       - Memory Current: 181.11 MB (Peak: 181.11 MB)
  TopMemoryTasksSnapshot:
    SnapshotManager@other_task@id=11480ad6017368b9-8aef2150427d229f:
       - Memory Current: 0 (Peak: 0)
    StreamLoadPipe@load@id=024958077445f4f7-d739fd5231dca890:
       - Memory Current: 1.39 GB (Peak: 1.39 GB)
    SegCompaction@compaction@id=2d4d5c796a9f48ee-fdf5dd9f8faf83a8:
       - Memory Current: 0 (Peak: 0)
    Load#Id=544468c7f5a9bac6-2c78fc9962f3908d@load@id=52497774c12c5ad1-1fda35667a3d0e8b:
       - Limit: 2.00 GB
       - Memory Current: 6.06 MB (Peak: 6.06 MB)
    Load#Id=624e7947099ef183-577a22cca9d34d89@load@id=ed4691ce5a748ca8-e49ef7cc9be10b95:
       - Limit: 2.00 GB
       - Memory Current: 6.06 MB (Peak: 6.06 MB)
 ```

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [x] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [x] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [x] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/runtime/memory/memory_profile.h [hunk 1]
```diff
@@ -21,6 +21,14 @@
 
 #include "util/runtime_profile.h"
 
+#define RETURN_MEMORY_PROFILE_STR(profile) \
```
Comment (yiguolei) [inline be/src/runtime/memory/memory_profile.h:24]: 不要使用macro，使用函数

File: be/src/olap/metadata_adder.h [hunk 2]
```diff
@@ -207,18 +213,24 @@ void MetadataAdder<T>::add_num(int64_t val) {
     }
     if constexpr (std::is_same_v<T, RowsetMeta>) {
         g_rowset_meta_num << val;
+        ExecEnv::GetInstance()->all_rowsets_mem_tracker()->consume(val);
```
Comment (yiguolei) [inline be/src/olap/metadata_adder.h:216]: 不要这么写。
重新写一个函数，每次调用tracker 更新的时候，把所有的bvar 的值往tracker 里更新一下。
这样在下面的代码里，我们就不需要记录tracker的这些值了，假如bvar 在不停的更新，我们的tracker 也就能跟着更新了

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 3]
```diff
@@ -200,6 +203,8 @@ Status Segment::_open() {
     _meta_mem_usage += (_num_rows + 1023) / 1024 * (36 + 4);
     // 0.01 comes from PrimaryKeyIndexBuilder::init
     _meta_mem_usage += BloomFilter::optimal_bit_num(_num_rows, 0.01) / 8;
+    _tracked_meta_mem_usage = _meta_mem_usage;
+    ExecEnv::GetInstance()->all_segments_mem_tracker()->consume(_tracked_meta_mem_usage);
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp:207]: 参考上面那个评论，不要改这里的代码了。

File: be/src/olap/segment_loader.h [hunk 4]
```diff
@@ -75,9 +75,15 @@ class SegmentCache : public LRUCachePolicy {
     // Holding all opened segments of a rowset.
     class CacheValue : public LRUCacheValueBase {
     public:
-        ~CacheValue() override { segment.reset(); }
+        CacheValue(segment_v2::SegmentSharedPtr segment_) : segment(std::move(segment_)) {
+            // Transfer to segment cache memory tracker.
+            ExecEnv::GetInstance()->all_segments_mem_tracker()->release(segment->meta_mem_usage());
```
Comment (yiguolei) [inline be/src/olap/segment_loader.h:80]: 这种逻辑改的太乱了，别这么改了，后面很难维护。
不如metadata adder 那里也像tablet meta 那种统计方式，把子对象累加起来就可以了。
然后cache 里继续使用meta_mem_usage 这个值。

File: be/src/olap/metadata_adder.h (hunk 2 above)
Comment (xinyiZzz) [reply be/src/olap/metadata_adder.h:216]: 2个目的，
一是实时性，tracker是要有一定实时性的，否则每次都要手动refresh，不能保证其他人用tracker会记得要refresh
二是维护性，如果单独搞一个方法refresh，把几个bvar的值相加，那其他人在 metadata_adder.h 新增元数据类的时候，不一定会想着同步修改refresh方法

File: be/src/olap/segment_loader.h (hunk 4 above)
Comment (xinyiZzz) [reply be/src/olap/segment_loader.h:80]: done

**Comment by yiguolei** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/common/daemon.cpp [hunk 5]
```diff
@@ -500,15 +500,17 @@ void Daemon::cache_adjust_capacity_thread() {
 void Daemon::cache_prune_stale_thread() {
     int32_t interval = config::cache_periodic_prune_stale_sweep_sec;
     while (!_stop_background_threads_latch.wait_for(std::chrono::seconds(interval))) {
-        if (interval <= 0) {
-            LOG(WARNING) << "config of cache clean interval is illegal: [" << interval
-                         << "], force set to 3600 ";
+        if (config::cache_periodic_prune_stale_sweep_sec <= 0) {
+            LOG(WARNING) << "config of cache clean interval is: [" << interval
+                         << "], cache prune stale thread pause 3600s ";
```
Comment (yiguolei) [inline be/src/common/daemon.cpp:505]: 我觉得原来的log，比现在你改的要好啊

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 6]
```diff
@@ -174,6 +178,17 @@ int64_t Segment::get_metadata_size() const {
            (_pk_index_meta ? _pk_index_meta->ByteSizeLong() : 0);
 }
 
+void Segment::update_metadata_size() {
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp:181]: 我感觉这个方法没必要重写，跟父类不一样的是188行。 
可能你再segment 里增加一个新的函数里面
{
MetaAdder::update_metadata_size();
g_segment_estimate_mem_bytes << _meta_mem_usage - _tracked_meta_mem_usage;
    _tracked_meta_mem_usage = _meta_mem_usage;

}

Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment.cpp:181]: done

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**